File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        
        ListNode current1 = list1;
        ListNode current2 = list2;
        ListNode mergeHead = new ListNode(0);
        ListNode mergeCurr = mergeHead;

        while(current1 != null && current2 != null) {
            if(current1.val < current2.val) {
                mergeCurr.next = current1;
                mergeCurr = mergeCurr.next;
                current1 = current1.next;
            } else {
                mergeCurr.next = current2;
                mergeCurr = mergeCurr.next;
                current2 = current2.next;
            }
        }

        if(current1 != null) {
            mergeCurr.next = current1;
        } else {
            mergeCurr.next = current2;
        }

        // while(current1 != null) {
        //     mergeCurr.next = current1;
        //     mergeCurr = mergeCurr.next;
        //     current1 = current1.next;
        // }
        
        // while(current2 != null) {
        //     mergeCurr.next = current2;
        //     mergeCurr = mergeCurr.next;
        //     current2 = current2.next;
        // }

        return mergeHead.next;
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        slow = fast = head
        # find mid point
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        # reverse second half
        second = slow.next
        slow.next = None # break the link to second half
        prev = None

        while second:
            temp = second.next
            second.next = prev
            prev = second
            second = temp

        # merge the two linked lists
        second = prev # reset second to true head
        first = head

        while second:
            temp1 = first.next
            temp2 = second.next

            first.next = second
            second.next = temp1

            first = temp1
            second = temp2

        